File: Data Structures and Algorithms GFG/Chapter-11 Matrix/Topic-3 Matrix Boundary Traversal/Program_1.java
/*Java Program to Print the Boundary Elements
of Fixed Sized 2D Array(Heap Allocation)*/

import java.util.*;

/*Creating an outer class*/
public class Program_1{

    /*Creating an inner class*/
    class Implement_Fixed_Sized_2D_Array{

        /*Data Members of class: int and int [][]*/
        int row_capacity, row_size, column_capacity, column_size;
        int [][]mat;

        /*Defining readInput Function*/
        void readInput(Scanner sc)
        {
            System.out.println("Enter the Row Capacity and Row Size of Matrix:");
            row_capacity=sc.nextInt();
            row_size=sc.nextInt();
            System.out.println("Enter the Column Capacity and Column Size of Matrix:");
            column_capacity=sc.nextInt();
            column_size=sc.nextInt();
        }

        /*Member Function of class:
        to Create a Fixed Sized 2D
        Array using Heap Allocation*/
        boolean create_a_fixed_sized_2D_array()
        {
            Scanner sc=new Scanner(System.in);
            readInput(sc);
            /*User is advised to enter a Non-Negative Integer that is greater than or 
            equal to 0 and less than (2 raised to power 31) as row and column capacity 
            and row and column size otherwise the result will not be as predicted*/

            /*Checking for bounds for row capacity and row size*/
            if(row_capacity>Integer.MAX_VALUE || row_size>Integer.MAX_VALUE || row_capacity<0 || row_size<0)
            {
                if(row_capacity>Integer.MAX_VALUE || row_capacity<0)
                {
                    System.out.println(row_capacity+" is an Invalid Input for Row Capacity!");
                    return false;
                }
                else 
                {
                    System.out.println(row_size+" is an Invalid Input for Row Size!");
                    return false;
                }
            }

            /*Checking for bounds for column capacity and column size*/
            if(column_capacity>Integer.MAX_VALUE || column_size>Integer.MAX_VALUE || column_capacity<0 || column_size<0)
            {
                if(column_capacity>Integer.MAX_VALUE || column_capacity<0)
                {
                    System.out.println(column_capacity+" is an Invalid Input for Column Capacity!");
                    return false;
                }
                else 
                {
                    System.out.println(column_size+" is an Invalid Input for Column Size!");
                    return false;
                }
            }

            /*Comparison of size and capacity*/
            if(row_size>row_capacity || column_size>column_capacity)
            {
                if(row_size>row_capacity)
                {
                    System.out.println("Row Size can't be greater than Row Capacity");
                    return false;
                }
                else
                {
                    System.out.println("Column Size can't be greater than Column Capacity");
                    return false;
                }
            }

            /*Allocating Memory to Matrix*/
            try{
                mat=new int[row_capacity][column_capacity];
            }
            catch(OutOfMemoryError e){
                System.out.println("Memory Allocation Failed!");
            }

            /*Taking the Elements of Matrix as User Input*/
            System.out.println("Enter "+(row_size*column_size)+" Elements for Matrix:");
            for(int i=0; i<row_size; i++)
            {
                for(int j=0; j<column_size; j++)
                {
                    System.out.print("Enter the Element for row "+(i+1)+" and column "+(j+1)+":");
                    int temp=sc.nextInt();
                    mat[i][j]=temp;
                }
            }
            sc.close();
            return true;
        }

        /*Member Function of class:
        to Display Elements of 2D Array:
        Display Function*/
        void display()
        {
            /*Displaying the Elements of Matrix: Traversing a Matrix*/
            System.out.println("The Elements of Matrix are:");
            for(int i=0; i<row_size; i++)
            {
                for(int j=0; j<column_size; j++)
                System.out.print(mat[i][j]+" ");
                System.out.println();
            }
        }

        /*Member Function of class:
        to Print Boundary Elements of 2D Array*/
        void print_boundary_elements()
        {
            /*Displaying the Boundary Elements of Matrix*/
            System.out.println("Boundary Elements of Matrix are:");
            if(row_size==1)
            {
                for(int i=0; i<column_size; i++)
                System.out.print(mat[0][i]+" ");
            }
            else if(column_size==1)
            {
                for(int i=0; i<row_size; i++)
                System.out.print(mat[i][0]+" ");
            }
            else
            {
                for(int i=0; i<column_size; i++)
                System.out.print(mat[0][i]+" ");
                for(int i=1; i<row_size; i++)
                System.out.print(mat[i][column_size-1]+" ");
                for(int i=column_size-2; i>=0; i--)
                System.out.print(mat[row_size-1][i]+" ");
                for(int i=row_size-2; i>=1; i--)
                System.out.print(mat[i][0]+" ");
            }
        }
    }

    /*Driver Function(Code)*/
    public static void main(String[] args)
    {
        Program_1 program_1=new Program_1();
        Program_1.Implement_Fixed_Sized_2D_Array IFS2A=program_1.new Implement_Fixed_Sized_2D_Array();
        if(!IFS2A.create_a_fixed_sized_2D_array())
        System.out.println("Matrix Creation Failed!");
        else
        {
            IFS2A.display();
            IFS2A.print_boundary_elements();
        }
    }
}

File: Data Structures and Algorithms GFG/Chapter-11 Matrix/Topic-2 Matrix in Snake Pattern/Program_1.cpp
/*C++ Program to Print the Elements of Fixed Sized 
2D Array(Heap Allocation) in Snake Pattern*/

#include<iostream>
#include<new>
#include<limits.h>
using namespace std;

/*Creating a class*/
class Implement_Fixed_Sized_2D_Array{
    public:

    /*Data Members of class:
    long long unsigned int and int** */
    long long unsigned int row_capacity, row_size, column_capacity, column_size;
    int **mat;

    /*Member Function of class:
    to Create a Fixed Sized
    2D Array using Heap Allocation*/
    bool create_a_fixed_sized_2D_array(void)
    {
        /*User Input for row capacity and row size*/
        cout<<"Enter the Row-Capacity and Row-Size of Matrix:";
        cin>>row_capacity>>row_size;
        /*User is advised to enter a Non-Negative Integer that is greater
        than or equal to 0 and less than (2 raised to power 64) as row capacity
        and row size otherwise the result will not be as predicted*/

        /*Checking for bounds for row capacity and row size*/
        if(row_capacity>ULLONG_MAX || row_size>ULLONG_MAX)
        {
            if(row_capacity>ULLONG_MAX)
            {
                cout<<row_capacity<<" is an Invalid Input for Row Capacity!"<<endl;
                return false;
            }
            else 
            {
                cout<<row_size<<" is an Invalid Input for Row Size!"<<endl;
                return false;
            }
        }

        /*User Input for column capacity and column size*/
        cout<<"Enter the Column-Capacity and Column-Size of Matrix:";
        cin>>column_capacity>>column_size;
        /*User is advised to enter a Non-Negative Integer that is greater
        than or equal to 0 and less than (2 raised to power 64) as column capacity
        and column size otherwise the result will not be as predicted*/

        /*Checking for bounds for column capacity and column size*/
        if(column_capacity>ULLONG_MAX || column_size>ULLONG_MAX)
        {
            if(column_capacity>ULLONG_MAX)
            {
                cout<<column_capacity<<" is an Invalid Input for Column Capacity!"<<endl;
                return false;
            }
            else 
            {
                cout<<column_size<<" is an Invalid Input for Column Size!"<<endl;
                return false;
            }
        }

        /*Comparison of size and capacity*/
        if(row_size>row_capacity || column_size>column_capacity)
        {
            if(row_size>row_capacity)
            {
                cout<<"Row Size can't be greater than Row Capacity"<<endl;
                return false;
            }
            else
            {
                cout<<"Column Size can't be greater than Column Capacity"<<endl;
                return false;
            }
        }

        /*Allocating Memory to Matrix*/
        mat=new(nothrow) int*[row_capacity];
        for(long long unsigned int i=0; i<row_capacity; i++)
        mat[i]=new(nothrow) int[column_capacity];

        /*Checking for Memory Allocation*/
        if(mat==NULL)
        {
            cout<<"Memory Allocation Failed!"<<endl;
            return false;
        }
        else 
        {
            for(long long unsigned int i=0; i<row_capacity; i++)
            {
                if(mat[i]==NULL)
                {
                    cout<<"Memory Allocation Failed!"<<endl;
                    return false;
                }
            }
        }

        /*Taking the Elements of Matrix as User Input*/
        cout<<"Enter "<<row_size*column_size<<" Elements for Matrix:"<<endl;
        for(long long unsigned int i=0; i<row_size; i++)
        {
            for(long long unsigned int j=0; j<column_size; j++)
            {
                cout<<"Enter the Element for row "<<(i+1)<<" and column "<<(j+1)<<":";
                cin>>mat[i][j];
            }
        }
        return true;
    }

    /*Member Function of class:
    to Display Elements of 2D Array:
    Display Function*/
    void display(void)
    {
        /*Displaying the Elements of Matrix: Traversing a Matrix*/
        cout<<"The Elements of Matrix are:"<<endl;
        for(long long unsigned int i=0; i<row_size; i++)
        {
            for(long long unsigned int j=0; j<column_size; j++)
            cout<<mat[i][j]<<" ";
            cout<<endl;
        }
    }

    /*Member Function of class:
    to Print Elements of 2D Array:
    in Snake Pattern*/
    void print_snake_pattern()
    {
        /*Displaying the Elements of Matrix in Snake Pattern*/
        cout<<"Elements of Matrix in Snake Pattern are:"<<endl;
        for(long long unsigned int i=0; i<row_size; i++)
        {
            if(i%2==0)
            {
                for(long long unsigned int j=0; j<column_size; j++)
                cout<<mat[i][j]<<" ";
            }
            else 
            {
                for(long long unsigned int j=column_size-1; j>=0; j--)
                {
                    if(j==0)
                    {
                        cout<<mat[i][j]<<" ";
                        break;
                    }
                    cout<<mat[i][j]<<" ";
                }
            }
        }
    }
};

/*Driver Function(Code)*/
int main()
{
    Implement_Fixed_Sized_2D_Array IFS2A;
    if(!IFS2A.create_a_fixed_sized_2D_array())
    cout<<"Matrix Creation Failed!";
    else
    {
        IFS2A.display();
        IFS2A.print_snake_pattern();
    }
    return 0;
}